agent/connect/testing_ca.go

```diff
@@ -57,13 +57,63 @@ func ValidateLeaf(caPEM string, leafPEM string, intermediatePEMs []string) error
 	return err
 }
 
-func testCA(t testing.T, xc *structs.CARoot, keyType string, keyBits int, ttl time.Duration) *structs.CARoot {
-	var result structs.CARoot
-	result.Active = true
+type TestCAOptions struct {
+	// KeyType for the private key, defaults to DefaultPrivateKeyType
+	KeyType string
+	// KeyBits for the private key, defaults to DefaultPrivateKeyBits
+	KeyBits int
+	// TTL for the root certificate, defaults to 10 years.
+	TTL time.Duration
+	// PreviousCARoot if not nil, will be used to cross-sign a certificate.
+	PreviousCARoot CertKeyPair
+}
+
+type TestCAResult struct {
+	Name         string
+	SigningKey   string
+	SigningKeyID string
+	RootCert     string
+	ID           string
+	SerialNumber uint64
+
+	CrossSigningCert string
+
+	// TODO:
+	//result.NotBefore = template.NotBefore.UTC()
+	//result.NotAfter = template.NotAfter.UTC()
+	//result.PrivateKeyType = opts.KeyType
+	//result.PrivateKeyBits = opts.KeyBits
+}
+
+func (r TestCAResult) KeyPair() CertKeyPair {
+	return CertKeyPair{Cert: r.RootCert, PrivateKey: r.SigningKey}
+}
+
+func (r TestCAResult) ToCARoot() *structs.CARoot {
+	// Note: not all fields are filled.
+	return &structs.CARoot{
+		ID:           r.ID,
+		SerialNumber: r.SerialNumber,
+		SigningKeyID: r.SigningKeyID,
+		RootCert:     r.RootCert,
+		Active:       true,
+	}
+}
+
+// NewTestCA creates a private key and CA certificate for testing.
+func NewTestCA(t testing.T, opts TestCAOptions) TestCAResult {
```

Comment:
This new function replaces the 3 existing ones that accepted different args but had very similar names. This allows us to add more options in the future without creating a new function every time. 

The sprawl of functions with similar names problem seems to be a common one. It makes code rigid and hard to change. This pattern of accepting a struct type as options, and returning a struct type as a result makes code much easier to change.

Instead of using a production type, we use a type specific to this helper. Using a production type (`CARoot`) is also a problem, because we need to make significant changes to that type (#11347) so overloading it with test-only fields makes that work more difficult.